cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -229,49 +231,24 @@ private void updateTetherStatus(HttpResponder responder, String peer, TetheringS
+   * Processes and publishes the list of tethering program updates received from the client
    */
-  private void processProgramUpdates(String content) {
-    if (content.isEmpty()) {
-      return;
+  private void processProgramUpdateNotifications(FullHttpRequest request) throws BadRequestException {
+    List<Notification> notificationList;
+    try (Reader reader = new InputStreamReader(new ByteBufInputStream(request.content()), StandardCharsets.UTF_8)) {
+      notificationList = GSON.fromJson(reader, NOTIFICATION_LIST_TYPE);
+    } catch (JsonSyntaxException | IOException e) {
+      throw new BadRequestException("Unable to parse request: " + e.getMessage(), e);
     }
-    TetheringProgramUpdate programUpdate = GSON.fromJson(content, TetheringProgramUpdate.class);
-    for (TetheringProgramUpdate.TetheringProgramUpdateRequest request : programUpdate.getUpdateRequests()) {
-      ProgramRunId receivedProgram = GSON.fromJson(request.getProgramRunId(), ProgramRunId.class);
-      ProgramRunId programRunId = new ProgramRunId(NamespaceId.SYSTEM.getNamespace(), receivedProgram.getApplication(),
-                                                   receivedProgram.getType(), receivedProgram.getProgram(),
-                                                   receivedProgram.getRun());
-      ProgramRunStatus programRunStatus = ProgramRunStatus.valueOf(request.getProgramStatus());
-      switch (programRunStatus) {
-        case RUNNING:
-          programStateWriter.running(programRunId, null);
-          break;
-        case STOPPING:
-          programStateWriter.stop(programRunId, 10);
-          break;
-        case COMPLETED:
-          programStateWriter.completed(programRunId);
-          break;
-        case KILLED:
-          programStateWriter.killed(programRunId);
-          break;
-        case FAILED:
-          programStateWriter.error(programRunId, null);
-          break;
-        case SUSPENDED:
-          programStateWriter.suspend(programRunId);
-          break;
-        case RESUMING:
-          programStateWriter.resume(programRunId);
-          break;
-        case REJECTED:
-          programStateWriter.reject(programRunId, null, null, null, null);
-          break;
-        default:
-          LOG.info("Unable to update programRunId {} to status {}", programRunId, programRunStatus);
-          return;
+
+    try {
+      for (Notification notification : notificationList) {
+        messagingContext.getMessagePublisher().publish(NamespaceId.SYSTEM.getNamespace(),
+                                                       cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC),
+                                                       GSON.toJson(notification));
       }
-      LOG.info("Updated programRunId {} to status {}", programRunId, programRunStatus);
+    } catch (TopicNotFoundException | IOException e) {
+      LOG.error("Unable to publish program status update");
```

Comment:
include the exception in log message `LOG.error("...", e)`

Reply:
fixed